### PR TITLE
🤏 New models for tests

### LINKED ISSUE #2177: Drop GPT2 in our test in favour of a more recent instruct model
Initially suggested here https://github.com/huggingface/trl/pull/2144#discussion_r1787905943

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes #2177

The models currently used for testing lack uniformity and practicality. Specifically:  
- They are sourced from various namespaces (`trl-internal-testing`, `facebook`, `philschmid`, `hf-internal-testing`, etc.).  
- They are not all small.  
- Their naming conventions are inconsistent (e.g., `tiny-random-MistralForCausalLM`, `dummy-GPT2-correct-vocab`, `gpt2`, `pythia-14m`).  
- There is no existing script for generating these models.  

This PR introduces the following improvements:  
- A script to create tiny models specifically for testing purposes.  
- A uniform naming convention for these models, all placed under a single namespace (`trl-internal-testing`).

When approved, I'll move every models into `trl-internal-testing` namespace (instead of `qgallouedec`). And probably remove all the old testing models in `trl-internal-testing`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/test_dataset_formatting.py [hunk 1]
```diff
@@ -136,9 +135,7 @@ def test_setup_chat_format(self):
         self.assertEqual(modified_tokenizer.eos_token, _chatml.eos_token)
         self.assertEqual(modified_tokenizer.pad_token, _chatml.pad_token)
         self.assertEqual(modified_tokenizer.bos_token, _chatml.bos_token)
-        self.assertEqual(len(modified_tokenizer), (original_tokenizer_len + 2))
```
Comment (qgallouedec) [inline tests/test_dataset_formatting.py:139]: no special token need to be added

File: tests/test_dataset_formatting.py [hunk 2]
```diff
@@ -136,9 +135,7 @@ def test_setup_chat_format(self):
...
         self.assertEqual((self.model.get_input_embeddings().weight.shape[0] % 64), 0)
-        self.assertEqual(self.model.get_input_embeddings().weight.shape[0], (original_tokenizer_len + 64))
```
Comment (qgallouedec) [inline tests/test_dataset_formatting.py:141]: Same here, no modification needed

File: tests/test_dpo_trainer.py [hunk 3]
```diff
@@ -659,36 +659,36 @@ def test_dpo_lora_bf16_autocast_llama(self):
...
+            ("robust", False, False),
+            ("robust", False, True),
+            ("robust", True, False),
+            ("robust", True, True),
         ]
     )
     @require_bitsandbytes
     @require_peft
     @unittest.skip("You need a GPU with bf16 support in order to run these tests")
-    def test_dpo_lora_bf16_autocast(self, name, loss_type, pre_compute, gen_during_eval):
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py:691]: not used

File: tests/test_dpo_trainer.py [hunk 4]
```diff
@@ -1185,9 +1186,8 @@ def test_vdpo_trainer(self, model_id):
             training_args = DPOConfig(
                 output_dir=tmp_dir,
                 per_device_train_batch_size=2,
-                max_length=512,
-                max_prompt_length=512,
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py:1189]: those param are optional now, so I remove them

File: tests/test_environments.py [hunk 5]
```diff
@@ -258,7 +258,7 @@ def test_text_environment_run(self, mock_generate):
         task_2 = "Hello there! General Kenobi!"
 
         query, response, response_mask, reward, histories = env.run([task_1, task_2])
-        self.assertEqual(len(query[0]), 9)
+        self.assertEqual(len(query[0]), 8)
```
Comment (qgallouedec) [inline tests/test_environments.py:261]: Le tokenizer a changé, de sorte que cette valeur est également

File: tests/test_kto_trainer.py [hunk 6]
```diff
@@ -156,9 +156,9 @@ def test_tokenize_and_process_tokens(self):
             self.assertListEqual(tokenized_dataset["prompt"], train_dataset["prompt"])
             self.assertListEqual(tokenized_dataset["completion"], train_dataset["completion"])
             self.assertListEqual(tokenized_dataset["label"], train_dataset["label"])
-            self.assertListEqual(tokenized_dataset["prompt_input_ids"][0], [5377, 11141])
-            self.assertListEqual(tokenized_dataset["prompt_attention_mask"][0], [1, 1])
-            self.assertListEqual(tokenized_dataset["answer_input_ids"][0], [318, 1365, 621, 8253, 13])
+            self.assertListEqual(tokenized_dataset["prompt_input_ids"][0], [31137])
+            self.assertListEqual(tokenized_dataset["prompt_attention_mask"][0], [1])
+            self.assertListEqual(tokenized_dataset["answer_input_ids"][0], [374, 2664, 1091, 16965, 13])
```
Comment (qgallouedec) [inline tests/test_kto_trainer.py:161]: The tokenizer has changed, so these value have also changed.

File: tests/test_kto_trainer.py [hunk 7]
```diff
@@ -196,15 +196,13 @@ def test_tokenize_and_process_tokens(self):
...
             self.assertListEqual(
-                processed_dataset["completion_input_ids"][0], [50256, 5377, 11141, 318, 1365, 621, 8253, 13, 50256]
-            )
-            self.assertListEqual(processed_dataset["completion_attention_mask"][0], [1, 1, 1, 1, 1, 1, 1, 1, 1])
-            self.assertListEqual(
-                processed_dataset["completion_labels"][0], [-100, -100, -100, 318, 1365, 621, 8253, 13, 50256]
+                processed_dataset["completion_input_ids"][0], [31137, 374, 2664, 1091, 16965, 13, 151645]
             )
+            self.assertListEqual(processed_dataset["completion_attention_mask"][0], [1, 1, 1, 1, 1, 1, 1])
+            self.assertListEqual(processed_dataset["completion_labels"][0], [-100, 374, 2664, 1091, 16965, 13, 151645])
```
Comment (qgallouedec) [inline tests/test_kto_trainer.py:205]: The tokenizer has changed, so this value has also changed.

File: tests/test_modeling_value_head.py [hunk 8]
```diff
@@ -24,38 +24,27 @@
...
-    "trl-internal-testing/tiny-random-MT5ForConditionalGeneration",
-    "trl-internal-testing/tiny-random-MvpForConditionalGeneration",
-    "trl-internal-testing/tiny-random-PegasusForConditionalGeneration",
-    "trl-internal-testing/tiny-random-PegasusXForConditionalGeneration",
-    "trl-internal-testing/tiny-random-PLBartForConditionalGeneration",
-    "trl-internal-testing/tiny-random-ProphetNetForConditionalGeneration",
-    "trl-internal-testing/tiny-random-SwitchTransformersForConditionalGeneration",
-    "trl-internal-testing/tiny-random-T5ForConditionalGeneration",
+    "qgallouedec/tiny-T5ForConditionalGeneration",
+    "qgallouedec/tiny-BartModel",
```
Comment (qgallouedec) [inline tests/test_modeling_value_head.py:47]: Only use the most popular two enc-dec models

File: tests/test_online_dpo_trainer.py [hunk 9]
```diff
@@ -32,16 +31,17 @@
 
 class TestOnlineDPOTrainer(unittest.TestCase):
     def setUp(self):
-        self.model_id = "trl-internal-testing/dummy-GPT2-correct-vocab"
+        self.model_id = "qgallouedec/tiny-Qwen2ForCausalLM-2.5"
         self.model = AutoModelForCausalLM.from_pretrained(self.model_id)
         self.ref_model = AutoModelForCausalLM.from_pretrained(self.model_id)
-        self.reward_model = AutoModelForSequenceClassification.from_pretrained("EleutherAI/pythia-14m", num_labels=1)
-        self.reward_tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-14m")
-        self.reward_tokenizer.chat_template = SIMPLE_CHAT_TEMPLATE
```
Comment (qgallouedec) [inline tests/test_online_dpo_trainer.py:40]: The model already have a chat template

File: tests/test_orpo_trainer.py [hunk 10]
```diff
@@ -21,25 +21,27 @@
...
         self.model = AutoModelForCausalLM.from_pretrained(self.model_id)
         self.tokenizer = AutoTokenizer.from_pretrained(self.model_id)
         self.tokenizer.pad_token = self.tokenizer.eos_token
 
         # get t5 as seq2seq example:
-        model_id = "trl-internal-testing/tiny-T5ForConditionalGeneration-correct-vocab"
+        model_id = "qgallouedec/tiny-T5ForConditionalGeneration"
         self.t5_model = AutoModelForSeq2SeqLM.from_pretrained(model_id)
         self.t5_tokenizer = AutoTokenizer.from_pretrained(model_id)
+        self.t5_tokenizer.chat_template = SIMPLE_CHAT_TEMPLATE
```
Comment (qgallouedec) [inline tests/test_orpo_trainer.py:38]: T5 doesn't have a chat template

File: tests/test_reward_trainer.py [hunk 11]
```diff
@@ -31,10 +31,10 @@
 
 class RewardTrainerTester(unittest.TestCase):
     def setUp(self):
-        self.model_id = "hf-internal-testing/tiny-random-LlamaForCausalLM"
+        self.model_id = "qgallouedec/tiny-Qwen2ForCausalLM-2.5"
         self.tokenizer = AutoTokenizer.from_pretrained(self.model_id)
-        self.tokenizer.chat_template = "{% for message in messages %}{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}{% if add_generation_prompt %}{{ '<|im_start|>assistant\n' }}{% endif %}"
```
Comment (qgallouedec) [inline tests/test_reward_trainer.py:36]: The model already have a chat template

File: tests/test_reward_trainer.py [hunk 12]
```diff
@@ -31,10 +31,10 @@
...
         self.model = AutoModelForSequenceClassification.from_pretrained(self.model_id)
+        self.model.config.pad_token_id = self.tokenizer.pad_token_id
```
Comment (qgallouedec) [inline tests/test_reward_trainer.py:37]: needed for batch forward in the model

File: tests/test_utils.py [hunk 13]
```diff
@@ -205,6 +205,7 @@ def setUp(self):
             ignore_index=self.ignore_index,
         )
 
+    @unittest.skip("This test must be updated.")
```
Comment (qgallouedec) [inline tests/test_utils.py:208]: This test seems very related to the way the codellama/CodeLlama-7b-Instruct-hf tokenizer works (especially on BOS and EOS) and is not general. It should be rewritten in the future.

File: tests/test_dpo_trainer.py [hunk 14]
```diff
@@ -1135,9 +1136,10 @@ def test_dpo_trainer_use_num_logits_to_keep(self):
 class DPOVisionTrainerTester(unittest.TestCase):
     @parameterized.expand(
         [
-            ["trl-internal-testing/tiny-random-idefics2"],
-            ["trl-internal-testing/tiny-random-paligemma"],
-            ["trl-internal-testing/tiny-random-llava-1.5"],
+            ("qgallouedec/tiny-Idefics2ForConditionalGeneration",),
+            # ("qgallouedec/tiny-PaliGemmaForConditionalGeneration",),
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py:1140]: paligemma doesn't have a chat template (not meant for chat). We've to find a way to test it. Dedicated method?

**Review by kashif — APPROVED** [review]

File: scripts/generate_tiny_models.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    if suffix is not None:
+        repo_id += f"-{suffix}"
+    model.push_to_hub(repo_id)
+    tokenizer.push_to_hub(repo_id)
+    model_card.push_to_hub(repo_id)
+
+
+# Decoder models
+for model_id, config_class, model_class, suffix in [
+    ("bigscience/bloomz-560m", BloomConfig, BloomForCausalLM, None),
```
[the remark this answers is not in the document]
Comment (lewtun) [reply scripts/generate_tiny_models.py:91]: Also, is this script supposed to be re-run whenever we add a model to the list? If so, I recommend adding a note either at the top of this script or in our contributor guide

**Review by lewtun — APPROVED** [review]
Excellent refactor and massive QoL improvements to the tests @qgallouedec! LGTM with a minor question about whether you want to distinguish between tiny base and instruct models

File: scripts/generate_tiny_models.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+---
+
+# Tiny {model_class_name}
+
+This is a minimal model built for unit tests in the [TRL](https://github.com/huggingface/trl) library.
+"""
+
+
+def push_to_hub(model, tokenizer, suffix=None):
+    model_class_name = model.__class__.__name__
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply scripts/generate_tiny_models.py:78]: Yes, I wasn't sure what to do about that. Most of our tests are based on Qwen2.5 in its instruct version. So I don't know how comptable the trainers are to non-instruct versions. Let's keep it like this for the moment.

File: scripts/generate_tiny_models.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply scripts/generate_tiny_models.py:102]: > of just relying on a small, curated set of popular architectures

Would you remove any model from this list?

> supposed to be re-run whenever we add a model

Yes. 

>  adding a note 

I added a note in the script in c851842

**Pull request merged**